File: day_07/p2.py
from typing import List, Tuple

def parse(fname: str) -> Tuple[ List[int], List[List[int]]]:
    lines: List[str] = []
    values = []
    operands = []

    with open(fname, 'r') as f:
        for line in f.read().splitlines():
            first, rest = line.split(':') 
            values.append(int(first))
            operands.append([int(op) for op in rest.split()])

    return values, operands

def perform(op1: int, op: str, op2: int) -> int:
    if op == '+':
        return op1 + op2
    if op == '*':
        return op1 * op2
    if op == '||':
        return int(str(op1) + str(op2))
    
    return -1

def dp(target: int, operands: List[int], operator: str) -> bool:
    # base case
    if len(operands) == 1:
        return operands[0] == target

    for operand in operands:
        new_operand = perform(operands[0], operator, operands[1])
        new_operands = operands[1:]
        new_operands[0] = new_operand
        operators = dp(target, new_operands, '+') or \
                    dp(target, new_operands, '*') or \
                    dp(target, new_operands, '||')

        return operators

    return False

def compute(values: List[int], operands: List[List[int]]):
    valid_eq_if_start_with_plus = set()
    valid_eq_if_start_with_mul = set()
    valid_eq_if_start_with_concat = set()

    for i, val in enumerate(values):
        if dp(val, operands[i], '+',):
            valid_eq_if_start_with_plus.add(i)
        if dp(val, operands[i], '*',):
            valid_eq_if_start_with_mul.add(i)
        if dp(val, operands[i], '||',):
            valid_eq_if_start_with_concat.add(i)

    valid = valid_eq_if_start_with_concat.union(
            valid_eq_if_start_with_plus.union(valid_eq_if_start_with_mul)
    )
    
    return sum(values[idx] for idx in valid)

def main():
    values, operands = parse("input.txt")
    print(compute(values, operands))


if __name__ == "__main__":
    main()

File: day_07/p2.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>

using namespace std;
using longs = vector<long long>;

pair<longs, vector<longs>> parse(const string& fname)
{
    longs values {};
    vector<longs> operands {};
    const auto lines = pypp::splitFileLines(fname);

    for (const auto& line : lines) {
        values.push_back(stoll(pypp::splitThenGetAt(line, ':', 0)));
        auto rest = pypp::splitThenGetAt(line, ':', 1);

        longs tmp {};
        for (const auto elem : pypp::split(rest, ' '))
            tmp.push_back(stoi(elem));

        operands.push_back(tmp);
    }

    return make_pair(values, operands);
}

long long perform(long long op1, string op, long long op2)
{
    if (op == "+")
        return op1 + op2;
    if (op == "*")
        return op1 * op2;
    if (op == "||")
        return stoll(to_string(op1) + to_string(op2));
}

bool dp(long long target, longs& operands, string op)
{
    // base case
    if (operands.size() == 1)
        return operands.front() == target;

    for (auto operand : operands) {
        vector<long long> new_operands {};
        if (op != "") {
            auto new_operand = perform(operands.at(0), op, operands.at(1));
            new_operands = vector<long long>(operands.begin() + 1, operands.end());
            new_operands.at(0) = new_operand;
        } else
            new_operands = operands;

        return dp(target, new_operands, "+") || dp(target, new_operands, "*") || dp(target, new_operands, "||");
    }

    return false;
}

long long compute(longs values, vector<longs>& operands)
{
    long long result { 0 };
    unordered_set<int> valid_indexes {};

    for (int i { 0 }; i < values.size(); ++i)
        if (dp(values[i], operands[i], ""))
            valid_indexes.insert(i);

    for (auto idx : valid_indexes)
        result += values.at(idx);

    return result;
}

int main()
{
    auto [values, operands] = parse("input.txt");
    std::cout << compute(values, operands);

    return 0;
}
